### DISCUSSION TITLE
cocotb testbench debug flow.

### DISCUSSION THREAD

**Discussion by Andrey7584** [cocotb/cocotb#3272] [Q&A]
I would like to debug testbench code written base on the cocotb framework. 
I am working in PyCharm, but unfortunately did not find a way to debug my testbench from PyCharm IDE.  Is current python cocotb testbench debug flow described in [cocotb-troubleshooting-debug](https://docs.cocotb.org/en/latest/troubleshooting.html?highlight=debug#python)? 
I tried setting PyCharm breakpoints in testbench  code and launching the simulation by "The Python Test Runner" in PyCharm but simulation did not stop on specified breakpoints. 
From cocotb documentation it is not clear whether it is possible to debug interactively cocotb testbnech from PyCharm IDE (or equivalent). 
In the makefile defined variable COCOTB_ATTACH. What type of debugger should be attached to this process?  C code debugger?

**Comment by marlonjames** [comment]
cocotb runs in an embedded Python interpreter inside the process of the simulator you are using.
If you have PyCharm professional license, you can use remote debugging: https://www.jetbrains.com/help/pycharm/remote-debugging-with-product.html
Instead of `remote_pdb` specified in the cocotb docs, you'll use `pydevd` or `pydevd-pycharm` package.

This answer to a previous discussion has an alternative approach:
https://github.com/cocotb/cocotb/discussions/2652#discussioncomment-1187266

**Comment by Andrey7584** [comment]
Thank you for your answer! Now it is clear how to debug the cocotb testbench python code.

Just one more thing. For  [COCOTB_ATTACH ](https://docs.cocotb.org/en/stable/building.html#envvar-COCOTB_ATTACH) makefile variable . "In order to give yourself time to attach a debugger to the simulator process before". What type of debugging/debugger COCOTB_ATTACH process used for?

**Reply by marlonjames** [reply]
That is primarily for gdb/lldb debuggers for the cocotb shared libraries.
I have attached a Python debugger and gdb at the same time for cocotb development.
